### ISSUE TITLE
State Decoder Training Issues

### ISSUE THREAD

**Issue by definitelyuncertain** [lmzintgraf/varibad#3]
Hi, I'm trying to get the state decoder working with the HalfCheetahDir-v0 task, but haven't had much success. When flipping on the state decoder with no other changes from the `args_mujoco_cheetah_dir_varibad.py` parameters, I get training curves like so:

![image](https://user-images.githubusercontent.com/20875718/82765953-b863ba00-9de0-11ea-96bb-7f701d28ce87.png)

As far as I can tell, introducing the state decoder shouldn't degrade the performance. If anything, it will have a regularizing effect on the decoder. So I'd just like to know if there are any recommended parameter settings that have been known to produce successful results, or if there's something I might be doing wrong.

I had to set the following arguments to get it to run:
```python
    parser.add_argument('--state_decoder_layers', nargs='+', type=int, default=[32, 32])
    parser.add_argument('--state_pred_type', type=str, default='deterministic',
                                        help='choose from: bernoulli, gaussian, deterministic')
    parser.add_argument('--state_loss_coeff', type=float, default=1.0, help='weight for state loss (vs reward loss)')
```

Also on this note, I could not get the Gaussian decoder to work. I get this error when I try to use it, and haven't been able to pinpoint what leads to this size mismatch:
```
  File "vae.py", line 168, in compute_state_reconstruction_loss
    loss_state = -m.log_prob(dec_next_obs).mean(dim=1)
  File "site-packages/torch/distributions/normal.py", line 76, in log_prob
    return -((value - self.loc) ** 2) / (2 * var) - log_scale - math.log(math.sqrt(2 * math.pi))
RuntimeError: The size of tensor a (21) must match the size of tensor b (42) at non-singleton dimension 2
```

**Comment by akuzeee** [comment]
Hi, I found a possible bug regarding the state decoder training. 
To insert 

```python
self.prev_obs[:, self.insert_idx] = self.running_prev_obs[i]
```

into [this line](https://github.com/lmzintgraf/varibad/blob/2fcca1cb73e863b5b91e66cb189c333afddde0dd/utils/storage_vae.py#L137) would hopefully solve the problem.

**Comment by lmzintgraf** [comment]
Thanks @akuzeee for pointing this out - that's fixed now in a36d251cd7abd10bf181b573ad7d9b86b9fa3787

@definitelyuncertain sorry for the delayed reply. After the bug fix above things should be better, but you might still see a slight performance drop when using the state decoder. Even though in principle the VAE should learn to just ignore the latent input (since it can predict the next state from the previous state + action alone), in practice it might not always do that. One thing to check is the magnitude of the reward decoder loss vs. the state decoder loss; it helps to adjust the state_loss_coefficient such that they're roughly of the same magnitude (otherwise, the gradients from the state decoder might overwhelm the ones from the reward decoder and the VAE can't learn).

I haven't actually tested the Gaussian state decoder with the latest code (and in all my experiments with it, it never actually helped performance), but let me know in case you still need this and I will look into it (without a multi-month delay this time :))
